IDinsight/ask-a-question
# [128] - Add GCS functionality to Existing Speech Workflow

Reviewer: @markbotterill 
Estimate: 20 mins

---

## Ticket

Fixes: [JIRA_TICKET_LINK](https://idinsight.atlassian.net/browse/AAQ-689?atlOrigin=eyJpIjoiNmZjNWNlMDNkMTVjNDM5OTlhZTk3MzYyNWU4NGY2YmIiLCJwIjoiaiJ9)

## Description
This PR integrates Google Cloud Storage (GCS) functionality into the `end-to-end speech workflow`, enabling cloud-based file `upload` and `retrieval`.

### Goal
the Goal of this PR is To enhance the existing speech-to-speech workflow by leveraging cloud storage, eliminating the need for local disk operations and facilitating easier file management.

### Changes
- Implemented two new functions in `core_backend/app/utils` for GCS file upload and retrieval.
- Integrated these functions into the `stt-llm-response` endpoint and `generate_speech` component.

## How has this been tested?
- dev environment
- docker compose
- Unit Tests

### How to test this?

1. Set the appropriate GCS variables from `template.core_backend.env` (according to your gcs bucket name, by default we using "aaq-speech-test")
2. Send a `POST` request to the` stt-llm-response` endpoint.
3. Verify the Uploaded file in the [GCS Bucket](https://console.cloud.google.com/storage/browser/aaq-speech-test%3Btab=objects?forceOnBucketsSortingFiltering=true&project=ask-a-question-421418&prefix=&forceOnObjectsSortingFiltering=false) .
4. If `generate_tts` is set to `true` in the QueryResponse Schema, a signed URL will be generated. Access this URL to play the generated MP3 file for text-to-speech output.

## Checklist

Fill with `x` for completed.
